### PR TITLE
Knownmethods optimizations

### PR DESCRIPTION
**Pull request by arespr**
**Before**
```
 |                  Method |       Mean |    StdDev |
 |------------------------ |----------- |---------- |
 |      GetKnownMethod_GET |  2.1957 ns | 0.0081 ns |
 |  GetKnownMethod_CONNECT | 10.5233 ns | 0.0800 ns |
 |   GetKnownMethod_DELETE |  3.4092 ns | 0.0148 ns |
 |     GetKnownMethod_HEAD |  2.5988 ns | 0.0079 ns |
 |    GetKnownMethod_PATCH |  2.5962 ns | 0.0120 ns |
 |     GetKnownMethod_POST |  2.6509 ns | 0.0857 ns |
 |      GetKnownMethod_PUT |  2.6619 ns | 0.0706 ns |
 |  GetKnownMethod_OPTIONS | 11.4744 ns | 0.0276 ns |
 |    GetKnownMethod_TRACE |  2.5957 ns | 0.0194 ns |
 | GetKnownVersion_HTTP1_1 |  2.7503 ns | 0.0160 ns |
```
**After**
```
 |                  Method |      Mean |    StdDev |
 |------------------------ |---------- |---------- |
 |      GetKnownMethod_GET | 2.1285 ns | 0.0037 ns |
 |  GetKnownMethod_CONNECT | 5.2002 ns | 0.0444 ns |
 |   GetKnownMethod_DELETE | 2.4482 ns | 0.0055 ns |
 |     GetKnownMethod_HEAD | 2.4525 ns | 0.0076 ns |
 |    GetKnownMethod_PATCH | 2.4550 ns | 0.0037 ns |
 |     GetKnownMethod_POST | 2.6373 ns | 0.1053 ns |
 |      GetKnownMethod_PUT | 2.4518 ns | 0.0155 ns |
 |  GetKnownMethod_OPTIONS | 5.2316 ns | 0.0109 ns |
 |    GetKnownMethod_TRACE | 2.4431 ns | 0.0035 ns |
 | GetKnownVersion_HTTP1_1 | 2.6444 ns | 0.0048 ns |
```

### CODE REVIEW

**Comment by davidfowl** [comment]
@arespr can you write a tiny description of the optimization?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs [hunk 1]
```diff
@@ -60,6 +61,33 @@ static HttpUtilities()
             _methodNames[(byte)HttpMethod.Options] = HttpMethods.Options;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetKnownMethodIndex(ulong value)
+        {
+            var tmp = (int)value & 0x100604;
+
+            return ((tmp >> 2) | (tmp >> 8) | (tmp >> 17)) & 0x0F;
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs]: Perfect hash? Did you write a program to generate the constants in this method? If so can you post it in the PR comments?

**Comment by arespr** [comment]
I have improved my search alghoritm and
found slightly better function:
```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static int GetKnownMethodIndex(ulong value)
{
   var tmp = (int) value & 0x0600000C;

    return ((tmp >> 2) | (tmp >> 23)) & 0x0F;
}
```
Alghoritm is here:
https://github.com/arespr/ConsoleApp1/blob/24ace01d17d1015c43c60d658319f1ed2fc5a043/src/ConsoleApp1/Utils.cs

**Comment by arespr** [comment]
added unit tests, and repaired performance tests

**Before**
```
 |                  Method |       Mean |    StdErr |    StdDev |     Median |
 |------------------------ |----------- |---------- |---------- |----------- |
 |      GetKnownMethod_GET |  2.0383 ns | 0.0021 ns | 0.0077 ns |  2.0379 ns |
 |  GetKnownMethod_CONNECT | 10.0297 ns | 0.1076 ns | 0.5486 ns |  9.7720 ns |
 |   GetKnownMethod_DELETE |  8.5029 ns | 0.0040 ns | 0.0144 ns |  8.5020 ns |
 |     GetKnownMethod_HEAD |  5.8423 ns | 0.0036 ns | 0.0140 ns |  5.8398 ns |
 |    GetKnownMethod_PATCH |  7.8644 ns | 0.0203 ns | 0.0761 ns |  7.8557 ns |
 |     GetKnownMethod_POST |  4.9119 ns | 0.0023 ns | 0.0087 ns |  4.9110 ns |
 |      GetKnownMethod_PUT |  4.0161 ns | 0.0044 ns | 0.0158 ns |  4.0174 ns |
 |  GetKnownMethod_OPTIONS | 10.6244 ns | 0.0125 ns | 0.0468 ns | 10.6132 ns |
 |    GetKnownMethod_TRACE |  6.8743 ns | 0.0102 ns | 0.0395 ns |  6.8686 ns |
 | GetKnownVersion_HTTP1_1 |  2.6691 ns | 0.0263 ns | 0.1417 ns |  2.5801 ns |
```
**After**
```
 |                  Method |      Mean |    StdErr |    StdDev |
 |------------------------ |---------- |---------- |---------- |
 |      GetKnownMethod_GET | 2.0009 ns | 0.0017 ns | 0.0066 ns |
 |  GetKnownMethod_CONNECT | 4.7313 ns | 0.0015 ns | 0.0057 ns |
 |   GetKnownMethod_DELETE | 5.9907 ns | 0.0596 ns | 0.3154 ns |
 |     GetKnownMethod_HEAD | 5.9507 ns | 0.0014 ns | 0.0053 ns |
 |    GetKnownMethod_PATCH | 4.9168 ns | 0.0489 ns | 0.2017 ns |
 |     GetKnownMethod_POST | 4.7472 ns | 0.0024 ns | 0.0094 ns |
 |      GetKnownMethod_PUT | 6.1486 ns | 0.0731 ns | 0.3014 ns |
 |  GetKnownMethod_OPTIONS | 4.7275 ns | 0.0013 ns | 0.0046 ns |
 |    GetKnownMethod_TRACE | 4.7980 ns | 0.0271 ns | 0.1049 ns |
 | GetKnownVersion_HTTP1_1 | 2.5617 ns | 0.0008 ns | 0.0030 ns |
```

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs [hunk 2]
```diff
@@ -60,6 +61,36 @@ static HttpUtilities()
...
+
+            return ((tmp >> 2) | (tmp >> 23)) & 0x0F;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void SetKnownMethod(ulong mask, ulong knownMethodUlong, HttpMethod knownMethod, int length)
+        {
+            _knownMethods[GetKnownMethodIndex(knownMethodUlong)] = new Tuple<ulong, ulong, HttpMethod, int, bool>(mask, knownMethodUlong, knownMethod, length, true);
+        }
+
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs]: nit: remove extra blank line

File: test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+            HttpMethod method;
+            int length;
+
+            bool result = data.GetKnownMethod(out method, out length);
+
+            Assert.Equal(expectedResult, result);
+            Assert.Equal(expectedMethod, method);
+            Assert.Equal(expectedLength, length);
+
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs]: nit: remove blank line

File: test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        [Theory]
+        [MemberData(nameof(TestData), MemberType = typeof(KnownStringsTests))]
+        public void GetsKnownMethod(byte[] methodData, HttpMethod expectedMethod, int expectedLength, bool expectedResult)
+        {
+            var data = new Span<byte>(methodData);
+
+            HttpMethod method;
+            int length;
+
+            bool result = data.GetKnownMethod(out method, out length);
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs]: nit: `var`

Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs]: and `out var` 😄

File: test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                CreateTestDataEntry(_methodGet, HttpMethod.Get, 3,true),
+                CreateTestDataEntry(_methodPut, HttpMethod.Put, 3,true),
+                CreateTestDataEntry(_methodPost, HttpMethod.Post, 4,true),
+                CreateTestDataEntry(_methodHead, HttpMethod.Head, 4,true),
+                CreateTestDataEntry(_methodTrace, HttpMethod.Trace, 5,true),
+                CreateTestDataEntry(_methodPatch, HttpMethod.Patch, 5,true),
+                CreateTestDataEntry(_methodDelete, HttpMethod.Delete, 6,true),
+                CreateTestDataEntry(_methodConnect, HttpMethod.Connect, 7,true),
+                CreateTestDataEntry(_methodOptions, HttpMethod.Options, 7,true),
+                CreateTestDataEntry(_invalidMethod1, HttpMethod.Custom, 0,false),
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs]: Why not just reuse the other arrays?

File: test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
+﻿using System;
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs]: Add copyright header.

**Comment by halter73** [comment]
I would like to know we can regenerate this if we change the list of known methods. Can we move the algorithm for https://github.com/arespr/ConsoleApp1/blob/24ace01d17d1015c43c60d658319f1ed2fc5a043/src/ConsoleApp1/Utils.cs to our CodeGenerator project, and have it generate the relevant code in HttpUtilities.cs? You can use the same conventions we use in KnownHeaders and FrameFeatureCollection (e.g. .Generated.cs partial class).

It's important that adding or modifying known methods is drop-in. So that probably involves at least generating the _knownMethods declaration so the array is the correct size, and of course generating GetKnownMethodIndex.

**Comment by arespr** [comment]
@halter73 I have added HttpUtiliies to CodeGenerator project.

**Comment by arespr** [comment]
**Dev**

```
 |                  Method |       Mean |    StdDev |
 |------------------------ |----------- |---------- |
 |      GetKnownMethod_GET |  2.1001 ns | 0.0115 ns |
 |  GetKnownMethod_CONNECT | 10.2516 ns | 0.0269 ns |
 |   GetKnownMethod_DELETE |  9.0024 ns | 0.3372 ns |
 |     GetKnownMethod_HEAD |  6.0905 ns | 0.0165 ns |
 |    GetKnownMethod_PATCH |  8.2099 ns | 0.1538 ns |
 |     GetKnownMethod_POST |  5.1370 ns | 0.0133 ns |
 |      GetKnownMethod_PUT |  4.1250 ns | 0.0271 ns |
 |  GetKnownMethod_OPTIONS | 11.3046 ns | 0.5302 ns |
 |    GetKnownMethod_TRACE |  7.1423 ns | 0.0229 ns |
 | GetKnownVersion_HTTP1_1 |  2.9267 ns | 0.0448 ns |
```

**Pr**

```
 |                  Method |      Mean |    StdDev |
 |------------------------ |---------- |---------- |
 |      GetKnownMethod_GET | 2.0916 ns | 0.0065 ns |
 |  GetKnownMethod_CONNECT | 5.0860 ns | 0.0183 ns |
 |   GetKnownMethod_DELETE | 5.0462 ns | 0.0240 ns |
 |     GetKnownMethod_HEAD | 4.9901 ns | 0.0426 ns |
 |    GetKnownMethod_PATCH | 5.0271 ns | 0.0287 ns |
 |     GetKnownMethod_POST | 5.0453 ns | 0.0296 ns |
 |      GetKnownMethod_PUT | 5.0553 ns | 0.0180 ns |
 |  GetKnownMethod_OPTIONS | 5.0340 ns | 0.0458 ns |
 |    GetKnownMethod_TRACE | 5.0139 ns | 0.0228 ns |
 | GetKnownVersion_HTTP1_1 | 2.8913 ns | 0.0071 ns |
```

File: test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        static byte[] _methodDelete = Encoding.ASCII.GetBytes("DELETE \0");
+        static byte[] _methodGet = Encoding.ASCII.GetBytes("GET ");
+        static byte[] _methodHead = Encoding.ASCII.GetBytes("HEAD \0\0\0");
+        static byte[] _methodPatch = Encoding.ASCII.GetBytes("PATCH \0\0");
+        static byte[] _methodPost = Encoding.ASCII.GetBytes("POST \0\0\0");
+        static byte[] _methodPut = Encoding.ASCII.GetBytes("PUT \0\0\0\0");
+        static byte[] _methodOptions = Encoding.ASCII.GetBytes("OPTIONS ");
+        static byte[] _methodTrace = Encoding.ASCII.GetBytes("TRACE \0\0");
+
+        const int MagicNumer = 0x0600000C;
```
Comment (Im5tu) [inline test/Microsoft.AspNetCore.Server.KestrelTests/KnownStringsTests.cs]: I assume this is a typo and you meant `MagicNumber`?

File: tools/CodeGenerator/HttpUtilities/HttpUtilities.cs [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+                    }
+
+                    tmpReturn += string.Format("(tmp >> {1})", HttpUtilitiesGeneratorHelpers.MaskToHexString(item.Mask), item.Shift);
+                }
+
+                var mask2 = (ulong)(Math.Pow(2, bitsCount) - 1);
+
+                string returnString = string.Format("return ({0}) & {1};", tmpReturn, HttpUtilitiesGeneratorHelpers.MaskToHexString(mask2));
+
+                bodyString = string.Format("            const int magicNumer = {0};\r\n            var tmp = (int)value & magicNumer;\r\n            {1}", HttpUtilitiesGeneratorHelpers.MaskToHexString(mask), returnString);
```
Comment (Im5tu) [inline tools/CodeGenerator/HttpUtilities/HttpUtilities.cs]: Same typo again here?

**Comment by arespr** [comment]
@Im5tu I can make it public and make shared code or remove id from unit tests.

**Comment by arespr** [comment]
@Im5tu or even leave as it is and it won't change test results.

**Comment by arespr** [comment]
@Im5tu  what do you propose ?

**Comment by halter73** [comment]
@arespr The only comments I see from @lm5tu are related to typo the "Numer" which should be changed "Number" instead in a couple places.

Thanks for making these changes. I'll review this more thoroughly Monday, and hopefully we can merge this.

File: tools/CodeGenerator/HttpUtilities/CombinationsWithoutRepetition.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            }
+
+            current[index - 1] = nElements[++pointers[index]];
+
+            return true;
+        }
+
+        private void ResetCurrent()
+        {
+            var p = _p;
```
Comment (davidfowl) [inline tools/CodeGenerator/HttpUtilities/CombinationsWithoutRepetition.cs]: Where is p used?

**Comment by halter73** [comment]
Thanks for your PR and your patience @arespr!

**Pull request merged**